ADR for allowing application code to be run-time initialized (#48427)
quarkusio/quarkus: adr/0009-native-class-initialization.adoc

classDef buildtime fill:#bbf;
class B,C,BT buildtime
....

=== Initialize all classes at build time

In this scenario, all classes in the Quarkus application are build time initialized, unless explicitly configured otherwise.
This option has been the default behavior in Quarkus for a long time.

[mermaid]
....
---
config:
  look: handDrawn
  theme: default
---
graph LR
  A[Quarkus Application] --> QA[io.quarkus:quarkus-arc]
  QA --> arc[io.quarkus.arc:arc] --> R1[...]
  arc --> jboss[org.jboss.logging:jboss-logging]
  QA --> core[io.quarkus:quarkus-core] --> R2[...]
  A --> UserExtension[com.example:my-extension] --> R3[...]
  UserExtension --> core
  UserExtension --> jboss
  A --> UserExtensionBT[com.example:build-time-initialized-extension] --> UserExtension
  A --> Camel[org.apache.camel.quarkus:camel-quarkus-core]  --> R4[...]
  A --> RestAssured[io.rest-assured:rest-assured:jar] --> R5[...]

classDef white fill:#fff;
class A white
classDef buildtime fill:#bbf;
class QA,arc,core,Camel,R1,R2,R3,R4,R5,UserExtension,UserExtensionBT,RestAssured,jboss buildtime
....

*Pros*:

- It allows Quarkus to take advantage of build time initialization as much as possible.

*Cons*:

> - It requires developers to deal with issues that arise from build time initialize and are often non-trivial.

Reviewer: ```suggestion
- It requires developers to deal with issues that arise from build time initialization that are often non-trivial.
```

- It requires developers to explicitly configure classes to be initialized at build time, which can be cumbersome.
- Is expected to result in bigger native executables.

=== Initialize some classes at build time and the rest at runtime

Since Quarkus already defaults to initializing classes at build time, it is possible to configure it to keep initializing some classes at build time while initializing the rest at runtime.
In this approach there are multiple scenarios between the two extremes of initializing all classes at build time or at runtime.
The more classes we initialize at build time, the more we can take advantage of build time initialization, but the more we risk running into issues that require additional configuration or code substitutions and vice versa.

==== Proposal 1: Initialize Quarkus core and all extensions along with their transitive dependencies at build time

In this scenario Quarkus core and all Quarkus extensions, including Quarkiverse extensions as well as user provided extensions along with their transitive closure of dependencies, are initialized at build time.

> This scenario reduces the entry barrier for most Quarkus users, as creating a custom Quarkus extensions is considered an advanced use case.

Reviewer: ```suggestion
This scenario reduces the entry barrier for most Quarkus users, as creating custom Quarkus extensions is considered an advanced use case.
```

[mermaid]
....
---
config:
  look: handDrawn
  theme: default
---
graph LR
  A[Quarkus Application] --> QA[io.quarkus:quarkus-arc]
  QA --> arc[io.quarkus.arc:arc] --> R1[...]
  arc --> jboss[org.jboss.logging:jboss-logging]
  QA --> core[io.quarkus:quarkus-core] --> R2[...]
  A --> UserExtension[com.example:my-extension] --> R3[...]
  UserExtension --> core
  UserExtension --> jboss
  A --> UserExtensionBT[com.example:build-time-initialized-extension] --> UserExtension
  A --> Camel[org.apache.camel.quarkus:camel-quarkus-core]  --> R4[...]
  A --> RestAssured[io.rest-assured:rest-assured:jar] --> R5[...]

classDef white fill:#fff;
class A white
classDef buildtime fill:#bbf;
class QA,arc,core,BT,Camel,R1,R2,R3,R4,UserExtension,UserExtensionBT,jboss buildtime
....

*Pros*:

- Changes the current behavior as little as possible.
- Reduces the entry barrier for most Quarkus users.

*Cons*:

- Keeps the entry barrier for extension developers high.

==== Proposal 2: Initialize Quarkus core and a vetted list of Quarkus extensions along with their transitive dependencies at build time

> In this scenario, Quarkus core and core extensions, as well as a vetted list of extensions along with the transitive closure of their dependencies, are initialized at build time, while user provided extensions and third-party libraries not handled by Quarkus are initialized at runtime.

Reviewer: ```suggestion
In this scenario, Quarkus core and core extensions, as well as a vetted list of extensions along with the transitive closure of their dependencies, are initialized at build time, while user provided extensions and third-party libraries are initialized at runtime.
```

[mermaid]
....
---
config:
  look: handDrawn
  theme: default
---
graph LR
  A[Quarkus Application] --> QA[io.quarkus:quarkus-arc]
  QA --> arc[io.quarkus.arc:arc] --> R1[...]
  arc --> jboss[org.jboss.logging:jboss-logging]
  QA --> core[io.quarkus:quarkus-core] --> R2[...]
  A --> UserExtension[com.example:my-extension] --> R3[...]
  UserExtension --> core
  UserExtension --> jboss
  A --> UserExtensionBT[com.example:build-time-initialized-extension] --> UserExtension
  A --> Camel[org.apache.camel.quarkus:camel-quarkus-core]  --> R4[...]
  A --> RestAssured[io.rest-assured:rest-assured:jar] --> R5[...]

classDef white fill:#fff;
class A white
classDef buildtime fill:#bbf;
class QA,arc,core,BT,Camel,R1,R2,R4,UserExtensionBT,jboss buildtime
....

*Pros*:

- Changes the current behavior as little as possible.
- Reduces the entry barrier for most Quarkus users as well as extension developers.

*Cons*:

> - Maintaining a vetted list of extensions might be cumbersome in the long run.

Reviewer: ```suggestion
- Maintaining a vetted list of extensions will be cumbersome in the long run.
```
I don't think there is a chance it won't be, especially when it's not dependencies from the core repository.

Quarkus applications may depend at run time on:

* Quarkus core and core extensions, which are distributed under the `io.quarkus` group ID.
* Quarkiverse extensions, which may be distributed under the `io.quarkiverse` or another group ID.
* User provided extensions not listed above
* Third-party libraries

The goal of this ADR is to define:
1. which of the above should be initialized at build time and which should be initialized at run time by default in native mode.
2. how developers can opt in or out of build time initialization.

== Decision Drivers

* Reduce entry barrier for developers that want to use Quarkus native mode
* Benefit from build time initialization when possible
* Allow developers to opt-in to build time initialization of parts that are run time initialized by default

== Considered options for default class initialization policy

[source, mermaid]
....
---
title: Legend
config:
  look: handDrawn
  theme: default
---
graph TD
  BT[Build-time initialized]
  RT[Run-time initialized]

classDef buildtime fill:#bbf;
class B,C,BT buildtime

> ....

Author: Yes, github renders these, see https://github.com/zakkak/quarkus/blob/2025-06-17-adr-build-init/adr/0009-native-class-initialization.adoc

=== Initialize all classes at build time

In this scenario, all classes in the Quarkus application are build time initialized, unless explicitly configured otherwise.
This option has been the default behavior in Quarkus for a long time.

[source, mermaid]
....
---
config:
  look: handDrawn
  theme: default
---
graph LR
  A[Quarkus Application] --> QA[io.quarkus:quarkus-arc]
  QA --> arc[io.quarkus.arc:arc] --> R1[...]
  arc --> jboss[org.jboss.logging:jboss-logging]
  QA --> core[io.quarkus:quarkus-core] --> R2[...]
  A --> UserExtension[com.example:my-extension] --> R3[...]
  UserExtension --> core
  UserExtension --> jboss
  A --> UserExtensionBT[com.example:build-time-initialized-extension] --> UserExtension
  A --> Camel[org.apache.camel.quarkus:camel-quarkus-core]  --> R4[...]
  A --> RestAssured[io.rest-assured:rest-assured:jar] --> R5[...]

classDef white fill:#fff;
class A white
classDef buildtime fill:#bbf;
class QA,arc,core,Camel,R1,R2,R3,R4,R5,UserExtension,UserExtensionBT,RestAssured,jboss buildtime
....

*Pros*:

- It allows Quarkus to take advantage of build time initialization as much as possible.

*Cons*:

> - It requires developers to deal with issues that arise from build time initialization that are often non-trivial.

Reviewer 2: When we talk about "developers", could we spell out if we're referring to our users vs the Quarkus code maintainers? The extension developers?  Sometimes the context makes it clear, but for example in this case I'm not sure what is being stated.

=== Initialize all classes at runtime

In this scenario, all classes in the Quarkus application are run time initialized (except for JDK classes footnote:[That's expected to change at some point after GraalVM for JDK 25]), unless explicitly configured otherwise.
This option has been the default behavior in GraalVM for a long time.

[source, mermaid]
....
---
config:
  look: handDrawn
  theme: default
---
graph LR
  A[Quarkus Application] --> QA[io.quarkus:quarkus-arc]
  QA --> arc[io.quarkus.arc:arc] --> R1[...]
  arc --> jboss[org.jboss.logging:jboss-logging]
  QA --> core[io.quarkus:quarkus-core] --> R2[...]
  A --> UserExtension[com.example:my-extension] --> R3[...]
  UserExtension --> core
  UserExtension --> jboss
  A --> UserExtensionBT[com.example:build-time-initialized-extension] --> UserExtension
  A --> Camel[org.apache.camel.quarkus:camel-quarkus-core]  --> R4[...]
  A --> RestAssured[io.rest-assured:rest-assured:jar] --> R5[...]

classDef white fill:#fff;
class A white
classDef buildtime fill:#bbf;
class UserExtensionBT buildtime
....

*Pros*:

> - It makes running code in native mode easier and more predictable as it behaves closer to how it would behave in JVM mode.

Reviewer 2: "easier" is relative. If you need to do a lof of _things_ which are vetoed within a native image, it's actually easier to run such code at build time.
e.g. introspection of a large model requires extensive reflection configurations. It's just _easier_ to do this at build time.
Even worse for bytecode manipulation/generation as there is no alternative:  and inherently this implies that any bytecode manipulation/generation work which depeds on some other task X would imply running X at build time.

Author: I will try to include that as well, but I don't think these are the common cases. So IMHO it's still easier to have things runtime initialized and only build-time initialize these exceptions than the opposite.

Reviewer 2: > I don't think these are common cases

Perhaps, but even if you say these are exceptional cases they exist and are the most interesting, and far tricker to analyse and address when they're not working as expected.

Remember, I'd never had been able to make Hibernate boot in native (as we had only some months to proof feasibility - I admit it could be done if we had unlimited time and people, but that's never the case) so Quarkus wouldn't have existed...  sure it can theoretically be done in different ways but it's extremely impractical.  We *can* rearchitecture things in this "easier" way, but it's not only a lot of engineering work to make this happen, it's also a lot of "work" that the JVM and our build tools will need to do at each build/compile/bootstrap.

So "easier" is not the right word here ..  as I said, it's relative :) Doesn't apply to certain extensions, in particular it doesn't apply to the most polished ones and I wish more extensions could optimise further for efficiency, there are several other key extensions which are known to be very heavy on reflective registrations leading to very bloated native-images and I suspect there could be better ways.

Author: Added https://github.com/quarkusio/quarkus/pull/48427/files#diff-2944ca1fe0897d5a84f747bfccb3241b19c2fe18ccc7c966c2db8cb180cb53aaR126

Author: > So "easier" is not the right word here .. as I said, it's relative :) Doesn't apply to certain extensions, in particular it doesn't apply to the most polished ones and I wish more extensions could optimise further for efficiency, there are several other key extensions which are known to be very heavy on reflective registrations leading to very bloated native-images and I suspect there could be better ways.

I agree and I believe we have a consensus that runtime-initialize everything is not the right thing to do. If you feel that the wording in the ADR as a whole (not just this bullet point) doesn't convey this please suggest possible amendments and I will merge them.

*Cons*:

- It does not allow Quarkus to take advantage of build time initialization, which can lead to slower startup times and higher memory usage.
- It requires developers to explicitly configure classes to be initialized at build time, which can be cumbersome.
- Is expected to result in bigger native executables.

=== Initialize some classes at build time and the rest at runtime

Since Quarkus already defaults to initializing classes at build time, it is possible to configure it to keep initializing some classes at build time while initializing the rest at runtime.
In this approach there are multiple scenarios between the two extremes of initializing all classes at build time or at runtime.
The more classes we initialize at build time, the more we can take advantage of build time initialization, but the more we risk running into issues that require additional configuration or code substitutions and vice versa.

==== Proposal 1: Initialize Quarkus core and all extensions along with their transitive dependencies at build time

> In this scenario Quarkus core and all Quarkus extensions, including Quarkiverse extensions as well as user provided extensions along with their transitive closure of dependencies, are initialized at build time.

Reviewer 2: Let's point out also that in several cases the "transitive closure of dependencies" might include code provided by the end-user, so even in this approach end-users still need to be somewhat careful and made aware of the rules.
(it's not a silver bullet solution)

Reviewer 2: Take Hibernate ORM for example; we need to at very least build the meta model at build time as that's the phase in which a very significant amount of introspection needs to be performed.
The user has plenty of opportunities to plug in custom strategies for many aspects of the model resolution, these are normally picked up via annotations but this isn't the only way.

For example, the table names in the databases are inferred from the names of the entity classes, and we'd really want to know the final table names during the metamodel generation. 
The user can:
 - not care, just let Hibernate infer the table name from the class name using a default strategy
 - override the table name on a case by case by using a `@Table(name=''@` annotation
 - plug in a [custom naming strategy ](https://docs.jboss.org/hibernate/orm/7.0/userguide/html_single/Hibernate_User_Guide.html#PhysicalNamingStrategy) - please have a look, it explains also why this is very common

I'd argue that any non-toy-demo application would apply a custom naming strategy (standard naming conventions are very common in any company of any size, are usually similar but never quite exactly the same so _some_ customization is necessary).

And this is just an example; I think we have hundreds of similar extension points.

From some more far-fetched perspective, the model itself is instrinsecally also a set of types that needs to be initialized at build time - entities are *normally* clear from odd code that would be concerning, but this isn't 100% guaranteed and our users *might* need to take this into account, again needing to learn facing some peculiarity of native images.

And this is jut Hibernate ORM. Jackson, Hibernate Validator, and many others all would allow some form of call-backs code into user code which needs to be triggered during the framework engine initialization.

Author: > Take Hibernate ORM for example; we need to at very least build the meta model at build time as that's the phase in which a very significant amount of introspection needs to be performed.
The user has plenty of opportunities to plug in custom strategies for many aspects of the model resolution, these are normally picked up via annotations but this isn't the only way.

Is this native-mode specific? I.e. in JVM-mode the meta model is built at run-time?

In general I think I get your point now. If I get this right, what you are saying is that Quarkus during the native image build due to build time initialization invokes the static initializers of some of its extensions which go through the application classpath and detect annotations or implementations of specific classes and (transitively) build time initialize them as well, even if they are user code.

I believe the best place to clarify this is in the level of `=== Initialize some classes at build time and the rest at runtime` as this applies to any mix of build-time with run-time initialization. WDYT?

Reviewer 2: > If I get this right, what you are saying is that Quarkus during the native image build due to build time initialization invokes the static initializers of some of its extensions which go through the application classpath and detect annotations or implementations of specific classes and (transitively) build time initialize them as well, even if they are user code.

Right, but beyond that: we don't merely initialize them, we execute the user code - with whatever other transitive dependencies that introduces, and consequences.

Advanced users creating custom extensions are expected to be able to configure their extensions to initialize classes at build time or run time as needed.

[source, mermaid]
....
---
config:
  look: handDrawn
  theme: default
---
graph LR
  A[Quarkus Application] --> QA[io.quarkus:quarkus-arc]
  QA --> arc[io.quarkus.arc:arc] --> R1[...]
  arc --> jboss[org.jboss.logging:jboss-logging]
  QA --> core[io.quarkus:quarkus-core] --> R2[...]
  A --> UserExtension[com.example:my-extension] --> R3[...]
  UserExtension --> core
  UserExtension --> jboss
  A --> UserExtensionBT[com.example:build-time-initialized-extension] --> UserExtension
  A --> Camel[org.apache.camel.quarkus:camel-quarkus-core]  --> R4[...]
  A --> RestAssured[io.rest-assured:rest-assured:jar] --> R5[...]

classDef white fill:#fff;
class A white
classDef buildtime fill:#bbf;
class QA,arc,core,BT,Camel,R1,R2,R3,R4,UserExtension,UserExtensionBT,jboss buildtime
....

*Pros*:

- Changes the current behavior as little as possible.
- Reduces the entry barrier for most Quarkus users.

*Cons*:

> - Keeps the entry barrier for extension developers high.

Reviewer 2: again please clarify which developers as I don't fully get it?

In this scenario, only Quarkus core and its transitive dependencies get initialized at build time.

[source, mermaid]
....
---
config:
  look: handDrawn
  theme: default
---
graph LR
  A[Quarkus Application] --> QA[io.quarkus:quarkus-arc]
  arc --> jboss[org.jboss.logging:jboss-logging]
  QA --> arc[io.quarkus.arc:arc] --> R1[...]
  QA --> core[io.quarkus:quarkus-core] --> R2[...]
  A --> UserExtension[com.example:my-extension] --> R3[...]
  UserExtension --> core
  UserExtension --> jboss
  A --> UserExtensionBT[com.example:build-time-initialized-extension] --> UserExtension
  A --> Camel[org.apache.camel.quarkus:camel-quarkus-core]  --> R4[...]
  A --> RestAssured[io.rest-assured:rest-assured:jar] --> R5[...]

classDef white fill:#fff;
class A white
classDef buildtime fill:#bbf;
class QA,arc,core,R1,R2,jboss buildtime
....

*Pros*:

- Reduces the entry barrier for most Quarkus users as well as extension developers.
- Reduce maintenance burden as only a small set of dependencies is initialized at build time.

*Cons*:

- Expected to result in slower startup times and higher memory usage.
- Expected to result in bigger native executables.

> - Expected to result in some behavior differences.

Reviewer 2: I also expect lower runtime performance, not just bootstrap times. We'll miss out on significant DCE opportunities, and that has wild combinatorial consequences.

Author: > I also expect lower runtime performance, not just bootstrap times. 

Correct.

> We'll miss out on significant DCE opportunities, and that has wild combinatorial consequences.

Well that's our feeling but we don't actually have numbers to back this up do we?

Reviewer 2: > Well that's our feeling but we don't actually have numbers to back this up do we?

You're right, we don't. Perhaps we should work on this, aiming for the GraalVM workshop as a target date to backup some discussions?

But thinking about this, it reminds me of another strong drawback which we should document here: I do clearly remember having situations in which the lack of DCO would lead to inclusion of code branches which would in turn lead to total compilation failures.   These were in the early days (things might have changed) and would involve typically codepaths in overriden methods, and having these DCO's properly then suggests an invocation path that would have been megamorphic can be simplified into a direct invocation.

Author: > You're right, we don't. Perhaps we should work on this, aiming for the GraalVM workshop as a target date to backup some discussions?

I agree it's worth exploring, but it's not trivial to do it in a realistic/representative manner with the current tooling. I tried building and running `picocli-native` which is pretty minimal in "runtime-initialize everything" mode and I managed to get it to build but then it fails at runtime which indicates that it will take some more effort to have a Quarkus build-time init vs Quarkus run-time init demo, especially for more complex apps/tests. I believe having the "opt-in", "opt-out" mechanisms in place that would allow us to easily switch big parts of code from build time init to runtime init might make such demos easier to conduct.

FWIW the changes I applied to Quarkus to enable it to build `picocli-native` with runtime-initialization are here  https://github.com/zakkak/quarkus/commit/502569a67fcb51c05acaf0e24251571ebada0bfa

Build time initialization of classes in native mode aligns with Quarkus' philosophy of doing as much as possible at build time to optimize runtime performance and reduce startup time.
However, build time initialization often leads to issues that require additional configuration or even code substitutions to be worked around.
This complicates the development process and discourages developers from using Quarkus' native mode due to increased complexity if they rely on libraries not handled by Quarkus or some of its extensions.

== Decision Drivers

* Reduce entry barrier for developers that want to use Quarkus native mode
* Benefit from build time initialization when possible

== Considered options

=== Initialize all classes at build time

This option has been the default behavior in Quarkus for a long time.
It allows Quarkus to take advantage of build time initialization as much as possible, but requires developers to deal with issues that arise from it.

=== Initialize all classes at runtime

This option has been the default behavior in GraalVM for a long time.
It makes running code in native mode easier and more predictable as it behaves closer to how it would behave in the JVM.

> However, it does not allow Quarkus to take advantage of build time initialization, which can lead to slower startup times and higher memory usage.

Author: > Do we have evidence of this?

Not strong evidence as in "I turn this flag on and off and I get these results", but we have good reasons (e.g. build time initialization prevents a significant amount of initialization code (static blocks) from running at run-time).

> If we could do that with Quarkus (runtime init everything) and show it upstream, they might be receptive to help improve.

That is indeed interesting, not only for upstream but for us to quantify how much difference this makes. Unfortunately, last time I checked doing an experiment like this required significant effort since a lot of the Quarkus code base is based on the assumption that everything is build time initialized by default. Perhaps we could do some more constraint experiments but I believe these would not be representative in the end. I will have another look to see if anything changed since my last try.

> However, it does not allow Quarkus to take advantage of build time initialization, which can lead to slower startup times and higher memory usage.

Author: I gave it another go and managed to get `picocli-native` to build (but it fails at runtime). I had to include the excluded `org.jboss.modules` dependency and explicitly build-time initialize some packages.

```

The results I am seeing (with this non representative test) are:

Key                                            | Value 1                                                     | Value 2                                                     | Difference (%)
-----------------------------------------------|----------------------------------------|-------------------------------------------------------------|------------------------------------
analysis_results.classes.jni                   | 61                                                          | 61                                                          | 0.00%         
analysis_results.classes.reachable             | 7686                                                        | 7954                                                        | 3.49%         
analysis_results.classes.reflection            | 2514                                                        | 2622                                                        | 4.30%         
analysis_results.classes.total                 | 9028                                                        | 9296                                                        | 2.97%         
analysis_results.fields.jni                    | 66                                                          | 66                                                          | 0.00%         
analysis_results.fields.reachable              | 10053                                                       | 10538                                                       | 4.82%         
analysis_results.fields.reflection             | 141                                                         | 145                                                         | 2.84%         
analysis_results.fields.total                  | 17960                                                       | 18619                                                       | 3.67%         
analysis_results.methods.foreign_downcalls     | -1                                                          | -1                                                          | -0.00%        
analysis_results.methods.jni                   | 55                                                          | 55                                                          | 0.00%         
analysis_results.methods.reachable             | 37649                                                       | 38800                                                       | 3.06%         
analysis_results.methods.reflection            | 2895                                                        | 2902                                                        | 0.24%         
analysis_results.methods.total                 | 65194                                                       | 67191                                                       | 3.06%         
analysis_results.types.jni                     | 61                                                          | 61                                                          | 0.00%         
analysis_results.types.reachable               | 7686                                                        | 7954                                                        | 3.49%         
analysis_results.types.reflection              | 2514                                                        | 2622                                                        | 4.30%         
analysis_results.types.total                   | 9028                                                        | 9296                                                        | 2.97%         
image_details.code_area.bytes                  | 17882928                                                    | 18615136                                                    | 4.09%         
image_details.code_area.compilation_units      | 23905                                                       | 24482                                                       | 2.41%         
image_details.image_heap.bytes                 | 20967424                                                    | 21172224                                                    | 0.98%         
image_details.image_heap.objects.count         | 243047                                                      | 246247                                                      | 1.32%         
image_details.image_heap.resources.bytes       | 158536                                                      | 159280                                                      | 0.47%         
image_details.image_heap.resources.count       | 57                                                          | 58                                                          | 1.75%         
image_details.total_bytes                      | 39268696                                                    | 40202944                                                    | 2.38%         
resource_usage.cpu.load                        | 15.949658135172749                                          | 16.120218579234972                                          | 1.07%         
resource_usage.cpu.parallelism                 | 32                                                          | 32                                                          | 0.00%         
resource_usage.cpu.total_cores                 | 32                                                          | 32                                                          | 0.00%         
resource_usage.garbage_collection.count        | 150                                                         | 211                                                         | 40.67%        
resource_usage.garbage_collection.max_heap     | 10970726400                                                 | 10812391424                                                 | -1.44%        
resource_usage.garbage_collection.total_secs   | 1.086                                                       | 1.24                                                        | 14.18%        
resource_usage.memory.peak_rss_bytes           | 1975046144                                                  | 1974968320                                                  | -0.00%        
resource_usage.memory.system_total             | 66497785856                                                 | 66497785856                                                 | 0.00%         
resource_usage.total_secs                      | 16.202356224                                                | 17.221306093                                                | 6.29%

> However, it does not allow Quarkus to take advantage of build time initialization, which can lead to slower startup times and higher memory usage.

Reviewer 3: I once again must say that we should always be setting `--allow-incomplete-classpath` (or whatever it was called). This is *exactly* how JVM mode works, so we're not losing anything by doing it, but it removes so many headaches around optional dependencies.

> However, it does not allow Quarkus to take advantage of build time initialization, which can lead to slower startup times and higher memory usage.

Author: @dmlloyd this is orthogonal to this issue though, it's being tracked here https://github.com/quarkusio/quarkus/issues/43409. Unless you imply that this ADR should address both issues I suggest we move the discussion there. I am quite interesting in understanding how:

> It could even improve analysis by increasing the discoverability of obviously dead code.

If you could provide an example or some more info it would be great.

Please see https://github.com/quarkusio/quarkus/issues/43409#issuecomment-2368558796 and https://github.com/quarkusio/quarkus/issues/43409#issuecomment-2983276599 for my opinion on the matter.

=== Initialize specific classes at build time and the rest at runtime

Since Quarkus already defaults to initializing classes at build time, it is possible to configure it to keep initializing these classes at build time while initializing the rest at runtime.

== Decision

The "Initialize specific classes at build time and the rest at runtime" option seems to be the best compromise between the two extremes.

> * Quarkus core, the core extensions, as well as Quarkiverse extensions will continue to initialize their classes at build time by default.

Reviewer 4: What might not be easy here is to determine if a library is runtime or build time.

For instance, dependencies of a Quarkus extension could be considered as build time but should we consider all extensions? Or only the vetted ones? Should we have a build item that allows to push the list of artifacts vetted for build time but it could turn out a giant maintenance nightmare?

How do you envision this?

Author: > Extensions today are expected to mark code as build time or runtime init anyways so main goal would be to enable users non- quarkus dependencies to work and for that we don't need high precision do we?

@maxandersen a potential issue with this approach (which why I understand @gsmet is bringing up) is the following. Assuming Quarkus (or an extension of it) has a build time dependency on library XYZ, XYZ is not going to be marked as reachable during the native-image build (since it a build time dependency) and thus it won't be initialized neither at build time nor at run-time. As a result, our regular testing won't catch any potential issues. Now if an application has a run-time dependency on XYZ it will get marked for build-time initialization (even though Quarkus doesn't explicitly support it) due to it being part of Quarkus' dependencies (even if a build-time dep). This might result in crashes/issues which is not any different from what we are facing today. Hopefully most issues of libraries falling in this category have already been solved and maintaining them won't be that bad.

Reviewer: Perhaps I'm misunderstanding, but in a given project we always know which dependency is used at runtime.

Reviewer 5: yeah - what @aloubyansky says - I'm assuming we would just limit us to mark the deps we know we rely on behaving certain way at runtime. 

If we wanted to be conservative we could expand it to build time deps too but that feels excessive somehow?
